### PR TITLE
User login implementation and Serialization

### PR DESCRIPTION
**Pull request by 1izardo**
# Description

This branch contains all of the changes necessary for the user login functionality. Note: this branch also includes the changes in #38.

# Summary of Changes

_Note: this includes changes from both #38 and #39._

## Database & Entity Stuff

- introduced a `UserDatabase` class to store `User` objects
	- this replaces `UserList`
- introduced a `Database` superclass
	- uses generics, so it supports the storing of `User`, `Player`, or `Team` entities
	- implements the `Serializable` interface for saving state
	- `UserDatabase`, `PlayerDatabase`, and `TeamDatabase` extend the `Database` superclass
		- they inherit the `addEntity`, `setEntities`, and `getEntities` methods
	- `TeamDatabase` implements it's own unique `updateRoster` method
		- TODO: add unique functionality to `PlayerDatabase` that makes it different from the base `Database` class (not necessary, but would be good)
- all `Database` classes are no longer static
	- instances must be created when the program first runs (in `main()`)
	- instances must be passed to whatever methods have to interact with the databases
		- an empty `PlayerDatabase` and `TeamDatabase` get passed to `CSVAdapter` so players can be added
		- (WIP) `Database`s must be passed to our search methods, and since those are called by our input methods, we must also pass them to the input methods
	- why did we do this? to make serialization possible, testing easier, and bugfixing less painful
- made `Player` and `Team` serializable
	- all `Player` subclasses inherit the interface as well
- added `updateTeam` method to `Player` (we will probably need this functionality later)
- created the enum `InputType` for our `InputBuilder` to use (rather than depending on strings)
	- also added the `LOGIN_DETAILS` input type for creating a login details input prompt
- the `InputData` interface now uses generics for accepting different databases
	- this is because our input classes will now require a `Database` to be passed (mentioned above)

## Login Stuff

- created a new `InputLogin` class which implements `InputData`
	- this prompts the user for a username and password, passes this information to the `LoginController`, and stores the login result and username as variables so `CommandLine` can decide how to proceed
	- much of `LoginController`'s logic has been extracted out into the `CommandLine` or `LoginUseCase`
		- TODO: `LoginController` may be unnecessary, we can probably just use `LoginUseCase` (which handles all of the login logic)
	- removed `LoginInputBoundary` (unneccesary and annoying)
- added `LoginPresenter` to present the user w/ messages that correspond to each login result (successful, wrong password, new user created)

## Serialization Stuff

- re-implemented `ReadWriter`
	- this is a now a general-purpose class to enable serialization, no longer just for users
	- it is also now a class instead of an interface (the interface wasn't really necessary since we only need one of these classes)
	- all the methods are static (it just makes things easier)
	- `loadDatabases` and `saveDatabases` do what their names suggest: loading in and creating/overwriting `.ser` files
- changes to `CSVAdapter`
	- now `TeamDatabase` handles adding teams better
		- we no longer need an accumulator for the teams added so far; all `CSVAdapter` has to do is call `updateRoster()`
		- if the team doesn't exist already, it will be created and added to the database
	- renamed `dataDump` to `processFile` for better accuracy
	- added new `processRow` method, which acts as a helper for `processFile`
		- it will look at each row in the file, add the player, and update the team
		- basically extracted the existing logic out of `dataDump`/`processFile`

## Command Line Stuff

- new `runPrompts` method gets called from `main`
	- each individual prompt has its own method (*e.g.* `runLoginPrompt`)
- the command line now supports saving and exiting
	- after each search prompt, they are asked if they want to continue or exit
	- `runContinuePrompt` handles this part

### CODE REVIEW

**Review by m-parvaneh — APPROVED** [review]
Everything looks good. Lots of necessary changes! 😈

**Review by sack6oy — APPROVED** [review]
I had great a lot of fun writing this code with you. :poggers:

**Review by Kaartik7 — APPROVED** [review]
Good progress :)

**Review by MichaelUmeh007 — APPROVED** [review]
